### PR TITLE
Let garbage collector send orphaning patches in parallel

### PR DESCRIPTION
**Pull request by caesarxuchao**
Garbage collector sends orphaning patches in parallel now.

### CODE REVIEW

**Comment by deads2k** [comment]
@caesarxuchao how many worker threads do we normally run in GC?

**Comment by deads2k** [comment]
> @caesarxuchao how many worker threads do we normally run in GC?

I just put together that if we have 5 worker threads and 10 qps, this doesn't by us a whole lot since they'll all end up bottlenecking anyway. :(

**Comment by caesarxuchao** [comment]
I think it depends on how heavy the GC workload is. In the case where orphaning is the only heavy gc workload in the entire cluster, which is the case of the e2e test, this PR is going to help. Because previously, one worker will notice the owner is deleted, and it will patch the dependents sequentially in one thread; now the worker will spawn goroutines to patch the dependents.

**Comment by deads2k** [comment]
The change looks straightforward to me.  Is there something left?

**Comment by caesarxuchao** [comment]
That's it. I updated the first comment.

File: pkg/controller/garbagecollector/garbagecollector.go [hunk 1]
```diff
@@ -409,17 +410,30 @@ func (gc *GarbageCollector) processDeletingDependentsItem(item *node) error {
...
+	for i := range dependents {
+		go func(dependent *node) {
+			defer wg.Done()
+			// the dependent.identity.UID is used as precondition
+			patch := deleteOwnerRefPatch(dependent.identity.UID, owner.UID)
+			_, err := gc.patchObject(dependent.identity, patch)
+			// note that if the target ownerReference doesn't exist in the
+			// dependent, strategic merge patch will NOT return an error.
+			if err != nil && !errors.IsNotFound(err) {
+				errCh <- fmt.Errorf("orphaning %s failed, %v", dependent.identity, err)
```
Comment (krmayankk) [inline pkg/controller/garbagecollector/garbagecollector.go]: nit: `orphaning for dependent %s failed`

**Comment by krmayankk** [comment]
@caesarxuchao did this fix a scale issue  ? was the orphaning serialized across resource types or within a single resource type ?

**Comment by caesarxuchao** [comment]
It was serialized across resource types. But because the clients have qps limits, I estimate this PR at most will double the throughput of orphaning operations in specific situations.

**Pull request merged**